arangod/RocksDBEngine/RocksDBEngine.cpp

```diff
@@ -1357,6 +1378,124 @@ RecoveryState RocksDBEngine::recoveryState() noexcept {
 TRI_voc_tick_t RocksDBEngine::recoveryTick() noexcept {
   return TRI_voc_tick_t(server().getFeature<RocksDBRecoveryManager>().recoveryTick());
 }
+  
+void RocksDBEngine::scheduleTreeRebuild(TRI_voc_tick_t database, std::string const& collection) {
+  MUTEX_LOCKER(locker, _rebuildCollectionsLock);
+  _rebuildCollections.emplace(std::make_pair(database, collection), /*started*/ false);
+}
+
+void RocksDBEngine::processTreeRebuilds() {
+  Scheduler* scheduler = arangodb::SchedulerFeature::SCHEDULER;
+  if (scheduler == nullptr) {
+    return;
+  }
+
+  uint64_t maxParallelRebuilds = 2;
+  uint64_t iterations = 0;
+  while (++iterations <= maxParallelRebuilds) {
+    if (server().isStopping()) {
+      // don't fire off more tree rebuilds while we are shutting down
+      return;
+    }
+
+    std::pair<TRI_voc_tick_t, std::string> candidate{};
+
+    {
+      MUTEX_LOCKER(locker, _rebuildCollectionsLock);
+      if (_rebuildCollections.empty() || 
+          _runningRebuilds >= maxParallelRebuilds) {
+        // nothing to do, or too much to do
+        return;
+      }
+
+      for (auto& it : _rebuildCollections) {
+        if (!it.second) {
+          // set to started
+          it.second = true;
+          candidate = it.first;
+          ++_runningRebuilds;
+          break;
+        }
+      }
+    }
+
+    if (candidate.first == 0 || candidate.second.empty()) {
+      return;
+    }
+    
+    bool queued = scheduler->queue(arangodb::RequestLane::CLIENT_SLOW, [this, candidate]() {
+      if (!server().isStopping()) {
+        TRI_vocbase_t* vocbase = nullptr;
+        try {
+          auto& df = server().getFeature<DatabaseFeature>();
+          vocbase = df.useDatabase(candidate.first);
+          if (vocbase != nullptr) {
+            auto collection = vocbase->lookupCollectionByUuid(candidate.second);
+            if (collection != nullptr && !collection->deleted()) {
+              LOG_TOPIC("b96bc", INFO, Logger::ENGINES)
+                  << "starting background rebuild of revision tree for collection " << collection->name();
+              Result res = static_cast<RocksDBCollection*>(collection->getPhysical())->rebuildRevisionTree();
```

Comment:
`rebuildRevisionTree` is effectively noexcept, because its internals are wrapped into a `catchToResut(...)`. So it should never throw, but in case of an exception return a `Result` with a failure. In case that the `Result` building throws an exception, we are screwed, but that can happy only in OOM case.